internal/generator/generator.go

```diff
@@ -0,0 +1,92 @@
+package generator
+
+import (
+	"fmt"
+
+	"github.com/rs/zerolog/log"
+
+	"github.com/nestoca/joy-generator/internal/gitrepo"
+	"github.com/nestoca/joy/api/v1alpha1"
+	"github.com/nestoca/joy/pkg/catalog"
+)
+
+type Generator struct {
+	repo *gitrepo.GitRepo
+}
+
+type Result struct {
+	// Release holds the release's values loaded from the yaml file in the catalog
+	Release *v1alpha1.Release `json:"release"`
+
+	// Environment holds the environment info where the release will be deployed. The full spec is not loaded to minimize the payload size
+	Environment *Environment `json:"environment"`
+
+	// RenderedValues is a yaml string that is the Release.spec.values rendered with any templated fields
+	RenderedValues string `json:"templatedValues"`
+}
+
+type Environment struct {
+	Name        string `json:"name"`
+	ClusterName string `json:"clusterName"`
+	Namespace   string `json:"namespace"`
+}
+
+func New(repo *gitrepo.GitRepo) *Generator {
+	return &Generator{
+		repo: repo,
+	}
+}
+
+// Run runs the generator and returns a slice of results. Each result contains the release, the environment where it
+// will be deployed and the rendered values string.
+func (r *Generator) Run() ([]*Result, error) {
+	// Make sure we have the latest catalog changes
+	err := r.repo.Pull()
+	if err != nil {
+		return nil, fmt.Errorf("pulling git repo: %w", err)
+	}
+
+	// Load Releases and relevant environment info (cluster name & namespace)
+	joyCatalog, err := catalog.Load(catalog.LoadOpts{
+		Dir:          r.repo.Directory(),
+		LoadEnvs:     true,
+		LoadReleases: true,
+		ResolveRefs:  true,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("loading catalog: %w", err)
+	}
+
+	var reconciledReleases []*Result
+	for _, releaseGroup := range joyCatalog.Releases.Items {
+		for _, release := range releaseGroup.Releases {
+			if release != nil {
+				log.Debug().Str("release", release.Name).Str("environment", release.Environment.Name).Msg("processing release")
+
+				renderedValues, err := RenderValues(release)
+				if err != nil {
+					log.Error().Err(err).Str("release", release.Name).Str("environment", release.Environment.Name).Msgf("error rendering values for release %s", release.Name)
+
+					// we don't want to fail the whole process if rendering one release fails, so we'll just skip this one
+					continue
```

Comment:
If we only log these errors and ignore/skip them, how will they be handled/known by ArgoCD?

Reply:
Unfortunately there is no mechanism to bubble errors up to ArgoCD itself. Even in the ApplicationSet controller, rendering errors are only shown in it's logs. 

To try and compensate for that, the structured logging includes a `"release"` and `"environment"` field so that errors can be filtered easily. Log-based alerting can also be configured to watch for these errors as well.

Reply:
Oh wow, that seems to me like an oversight in their design... 😕

Reply:
Kinda, but also theres nowhere to bubble up the error to. Resource that "triggers" the reconciliation is the ApplicationSet, which is not visible anywhere in the UI. Maybe they'll find a better way in a future release.

Reply:
Ah, I get it.  ApplicationSets are not truly first-class, long-lived citizens.  They only generate the Applications, which are what devs see and manage via the UI.  That's definitely a shortcoming of the design.  ApplicationSets should be surfaced and tracked, because they _will_ fail in many cases when people change and break stuff.  If the ApplicationSets are neither managed in a pipeline nor tracked actively / displayed in UI, we will definitely need to collect those errors manually and set up our own alerting.